lib/common/tests/rules/pcmk__evaluate_date_expression_test.c

```diff
@@ -0,0 +1,1061 @@
+    crm_time_t *t = crm_time_new("2024-02-01");
+
+    assert_int_equal(pcmk__evaluate_date_expression(NULL, NULL, NULL), EINVAL);
+    assert_int_equal(pcmk__evaluate_date_expression(xml, NULL, NULL), EINVAL);
+    assert_int_equal(pcmk__evaluate_date_expression(NULL, t, NULL), EINVAL);
+
+    crm_time_free(t);
+    free_xml(xml);
+}
+
+static void
+null_next_change_ok(void **state)
+{
+    xmlNodePtr xml = string2xml(EXPR_LT_VALID);
+    crm_time_t *t = crm_time_new("2024-01-01");
+
+    assert_int_equal(pcmk__evaluate_date_expression(xml, t, NULL),
+                     pcmk_rc_within_range);
+
+    crm_time_free(t);
+    free_xml(xml);
+}
+
+#define EXPR_ID_MISSING                             \
+    "<" PCMK_XE_DATE_EXPRESSION " "                 \
+        PCMK_XA_OPERATION "='" PCMK_VALUE_LT "' "   \
+        PCMK_XA_END "='2024-02-01 15:00:00' />"
+
+static void
+id_missing(void **state)
+{
+    // Currently acceptable
+    xmlNodePtr xml = string2xml(EXPR_ID_MISSING);
+    crm_time_t *t = crm_time_new("2024-01-01");
+
+    assert_int_equal(pcmk__evaluate_date_expression(xml, t, NULL),
+                     pcmk_rc_within_range);
+
+    crm_time_free(t);
+    free_xml(xml);
+}
+
+#define EXPR_OP_INVALID                                 \
+    "<" PCMK_XE_DATE_EXPRESSION " " PCMK_XA_ID "='e' "  \
+        PCMK_XA_OPERATION "='not-a-choice' />"
+
+static void
+op_invalid(void **state)
+{
+    xmlNodePtr xml = string2xml(EXPR_OP_INVALID);
+    crm_time_t *t = crm_time_new("2024-01-01");
+
+    assert_int_equal(pcmk__evaluate_date_expression(xml, t, NULL),
+                     pcmk_rc_undetermined);
+
+    crm_time_free(t);
+    free_xml(xml);
+}
+
+#define EXPR_LT_MISSING_END                             \
+    "<" PCMK_XE_DATE_EXPRESSION " " PCMK_XA_ID "='e' "  \
+        PCMK_XA_OPERATION "='" PCMK_VALUE_LT "' />"
+
+static void
+lt_missing_end(void **state)
+{
+    xmlNodePtr xml = string2xml(EXPR_LT_MISSING_END);
+    crm_time_t *t = crm_time_new("2024-01-01");
+
+    assert_int_equal(pcmk__evaluate_date_expression(xml, t, NULL),
+                     pcmk_rc_undetermined);
+
+    crm_time_free(t);
+    free_xml(xml);
+}
+
+#define EXPR_LT_INVALID_END                             \
+    "<" PCMK_XE_DATE_EXPRESSION " " PCMK_XA_ID "='e' "  \
+    PCMK_XA_OPERATION "='" PCMK_VALUE_LT "' "           \
+    PCMK_XA_END "='not-a-datetime' />"
+
+static void
+lt_invalid_end(void **state)
+{
+    xmlNodePtr xml = string2xml(EXPR_LT_INVALID_END);
+    crm_time_t *t = crm_time_new("2024-01-01");
+
+    assert_int_equal(pcmk__evaluate_date_expression(xml, t, NULL),
+                     pcmk_rc_undetermined);
+
+    crm_time_free(t);
+    free_xml(xml);
+}
+
+static void
+lt_valid(void **state)
+{
+    xmlNodePtr xml = string2xml(EXPR_LT_VALID);
+    crm_time_t *t = NULL;
+    crm_time_t *next_change = NULL;
+    crm_time_t *reference = NULL;
+
+    // t is before end, next change is before end
+    t = crm_time_new("2023-01-01 05:00:00");
```

Comment:
All the tests where everything is valid could be calls to a five-argument function, if you want:
```
static [inline] void
func(xmlNode *xml, const char *now_s, const char *next_change_s,
     const char *reference_s, int rc)
{
    crm_time_t *now = crm_time_new(now_s);
    crm_time_t *next_change = crm_time_new(next_change_s);
    crm_time_t *reference = crm_time_new(reference_s);
    
    assert_int_equal(pcmk__evaluate_date_expression(xml, now, next_change), rc);
    assert_int_equal(crm_time_compare(next_change, reference), 0);
    crm_time_free(reference);
    crm_time_free(next_change);
    crm_time_free(t);
}
```

E.g.,
```
func(xml, "2024-02-01 14:59:59", "2024-02-01 18:00:00", "2024-02-01 15:00:00",
     pcmk_rc_within_range);
```

Reply:
Not worth it -- for 3.0.0, all except one will become invalid syntax errors

Reply:
I mean:
* `lt_valid()` has four sequences of nine lines that could each become one two-line function call as described above
* `gt_valid()` has five of them
* `range_valid_start_only()` has four of them
* `range_valid_end_only()` has four of them
* `range_valid_start_valid_end()` has six of them
* `range_valid_start_valid_duration()` has six of them

You certainly don't have to write the helper function, but these are all tests for *all valid* arguments. These aren't tests that will go away or be condensed in 3.0.0.

Reply:
Ah I see what you mean. Makes sense.